### PR TITLE
Rewind dir fd before using it for cleanup

### PR DESCRIPTION
**Pull request by keszybz**
@DaanDeMeyer you said that you reproduced the issue in #29606. Could you test if this solution also resolves it?

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
@keszybz There's a very trivial reproducer included in https://github.com/systemd/systemd/issues/29606 that will work on Fedora 38 and Fedora 39. I say you if you change it, you're on the hook for verifying it as well ;).

File: src/shared/copy.c [hunk 1]
```diff
@@ -638,6 +638,10 @@ static void hardlink_context_destroy(HardlinkContext *c) {
          * _cleanup_() so that we really delete this, even on failure. */
 
         if (c->dir_fd >= 0) {
+                /* <dir_fd> might be have already been used for reading, so we need to rewind it. */
```
Comment (bluca) [inline src/shared/copy.c:641]: This is obscure enough that I think it deserves a comment in src/shared/rm-rf.h to warn future users to be careful about this

**Review by bluca — APPROVED** [review]

**Comment by keszybz** [comment]
Testing confirms that this also works. (I only asked because I though you have the reproducer fresh in mind and can test it with minimal effort. I didn't know that the reproducer is so reliable and quick to execute.)

Pushed with an additional comment and adjusted commit message.

**Review by DaanDeMeyer — APPROVED** [review]

**Comment by keszybz** [comment]
CentosCI (Arch Linux) failed in TEST-50-DISSECT, i.e. #28141.

**Pull request merged**